server/src/main/java/org/elasticsearch/action/search/SearchContextId.java

```diff
@@ -36,17 +38,27 @@ public final class SearchContextId {
     private final Map<ShardId, SearchContextIdForNode> shards;
     private final Map<String, AliasFilter> aliasFilter;
     private final transient Set<ShardSearchContextId> contextIds;
+    private final Map<SearchShardTarget, String> failedShardsWithCauses;
 
-    SearchContextId(Map<ShardId, SearchContextIdForNode> shards, Map<String, AliasFilter> aliasFilter) {
+    SearchContextId(
+        Map<ShardId, SearchContextIdForNode> shards,
+        Map<String, AliasFilter> aliasFilter,
+        Map<SearchShardTarget, String> failedShards
```

Comment:
I think we can simplify this a bit.  From a high-level perspective, we want to report a shard missing exception in the search response. The original reason for a shard not being included in the pit isn't valuable for the user. Since the pit is static, the user can either recreate it or acknowledge that some data is missing.

So, can we add a null entry in the shards map for the missing ShardId? This way, we can create a SearchShardIterator with an empty targetNodeIds list, and the logic for handling missing shards will take care of the rest. 
However, we need to ensure backward compatibility, in case a pit created in a newer version is used on older nodes. We can consider ignoring these missing shards in such case and document it since using the new allow_partial_search_result parameter shouldn't be used in a mixed cluster?

Reply:
Changes pushed in https://github.com/elastic/elasticsearch/pull/111516/commits/2f6addc7d323b51f1504cc3904ccde0825ed3479